docs/types.md

```diff
@@ -0,0 +1,187 @@
+
+```rust
+// This is almost it, but not quite
+fn largest<T>(list: &[T]) -> T {
+    let mut largest = list[0];
+
+    for &item in list {
+        if item > largest {
+            largest = item;
+        }
+    }
+    largest
+}
+```
+
+There are two lines of code that cause problems:
+
+```rust
+        // The use of the `>` operator is problem #1
+        if item > largest {
+            // The use of the `=` operator between two non-reference types is problem #2
+            largest = item;
+```
+
+Simply put, I need generic types that have the `PartialOrd` and `Copy` traits to solve problems #1 and #2 respectively. As a result, the final iteration of this method looks like the following:
+
+```rust
+// This is almost it, but not quite
+fn largest<T: PartialOrd + Copy>(list: &[T]) -> T {
+    let mut largest = list[0];
+
+    for &item in list {
+        if item > largest {
+            largest = item;
+        }
+    }
+    largest
+}
+```
+
+Note how there is room for a list of possible generic types between the function name and parameter list; we'll be working in that area a bit more in the next section.
+
+In terms of performance concerns, Rust has got you covered. Instead of checking for compliance of types with the generics at runtime, Rust makes checking compliance with generic specifiers a job for the compiler. Furthermore, the compiler will compile a list of all types that use said generic, and will generate copies of the generic function for each specific type. For example, the following code:
+
+```rust
+let int_list = vec![1, 2, 3];
+let largest_int = largest(int_list);
+let float_list = vec![1.0, 2.0, 3.0];
+let largest_float = largest(float_list);
+```
+
+Will end up having the following functions after compiling:
+
+```rust
+fn largest_i32(list: &[i32]) -> i32 { ... }
+fn largest_f32(list: &[f32]) -> f32 { ... }
+```
+
+As such, Rust gives up a bit of space in order to not have any performance cost for using generics.
+
+## lifetimes
+
+One large problem that has yet to be addressed in our exploration of Rust is that of returning one of many references. Consider the following function that returns the shorter of two string _references_:
+
+```rust
+// This won't compile
+fn shorter_string(str1: &str, str2: &str) -> &str {
+    if (str1.len() < str2.len()) {
+        str1
+    } else {
+        str2
+    }
+}
+```
+
+The problem is that Rust doesn't know what lifetime to assign to the return value of the function. You might think that it should just automatically give everything the same lifetime, but that won't always work. This is perhaps best displayed in the following snippet of code:
+
+```rust
+// This won't compile either
+fn main() {
+    let str1 = String.from("blah blah blah");
+    {
+        let str2 = String.from("ha ha ha ha");
+        let shortest = shorter_string(str1.as_str(), str2.as_str());
+    }
+    println!("the shorter string is {}", shortest);
```

Comment:
I feel like this bit of code is no longer showing what the piece of code from the Rust book shows, as shortest just goes out of scope without even thinking about str1 or str2. Shouldn't you declare shortest in the same scope as str1 like the below code?
```
fn main() {
    let string1 = String::from("long string is long");
    let result;
    {
        let string2 = String::from("xyz");
        result = longest(string1.as_str(), string2.as_str());
    }
    println!("The longest string is {}", result);
}
```

Reply:
Good catch; I'll get that fixed.